### ISSUE TITLE
font overwrite not working on 2.0.5

### ISSUE THREAD

**Issue by arbymou** [cleidigh/printing-tools-ng#69]
after updating the font overwrite stopped working. when i downgraded back to 2.0.4 it worked again

**Comment by cleidigh** [comment]
@arbymou 
thanks for the post.
some clarification:
when you say 'font overwrite'  do you mean One of the two font settings in the style tab?
if it is the content font Can you  send me an example e-mail and image of the problem?
test1@kokkini.net
thanks
Christopher

**Comment by kiki-ja** [comment]
@arbymou 
@cleidigh 

This is when select a font for the message body or header string in the style tab.
When I change the font in PrintingTools NG 2.0.5, it is not reflected in the print preview and print out.(no ovewritten)

The font settings are as specified in the Thunderbird options.
The same result is obtained with PrintingTools NG 2.0.4 and older versions.

I attach screenshot and printed out PDF file.
In the screenshot, I chose the 'Impact' font.
The body of the message was not changed to the 'Impact' font.

OS: Windows 10 Pro 64bit 20H2
Thunderbird 78.10.1 64bit

[Print Out Sample 2.0.5.pdf](https://github.com/cleidigh/printing-tools-ng/files/6458774/Print.Out.Sample.2.0.5.pdf)

![Font Select Sample 2 0 5](https://user-images.githubusercontent.com/80428201/117807195-42c55800-b296-11eb-92e7-d63499264019.png)

**Comment by cleidigh** [comment]
@kiki-ja 
so if I understand you correctly the fonts to not work in the older versions at all?
I thought you and others used fonts successfully?
Christopher

**Comment by kiki-ja** [comment]
@cleidigh 

Yes, it is not working for all versions 2.0.0, 2.0.1, 2.0.2, 2.0.4, and 2.0.5.
Not working for both plain text messages and HTML text messages.
However, font size change is working.

I understood from the beginning that it was not working.
I'll continue to test and confirm.

**Comment by kiki-ja** [comment]
@arbymou 
@cleidigh 

I'm sorry. There was an error on my part.

The font change worked in older versions.
However, in version 2.0.5, the following conditions were required.

1. Thunderbird 'Option' -> 'General' -> 'Language & Appearance' -> 'Fonts & colors' -> 'Advanced' -> 'Fonts & Encodings' open.
2. 'Font Control' -> Allow messages to use other fonts > checked.

See the screenshot.

I had the above unchecked, so the font change did not work.
It worked when I checked.

**Options - Language & Appearance - Advanced**
![Options - Language   Appearance - Advanced](https://user-images.githubusercontent.com/80428201/118198277-87feab00-b48b-11eb-8784-968bb64081c2.png)

**Fonts & Encodings**
![Fonts   Encodings](https://user-images.githubusercontent.com/80428201/118198289-8d5bf580-b48b-11eb-8837-aa9e5600d199.png)

**Font Change Worked**
![Font Change Worked](https://user-images.githubusercontent.com/80428201/118198307-90ef7c80-b48b-11eb-94c9-dcec49b21837.png)

**Comment by cleidigh** [comment]
@kiki-ja 
Thanks ! this should be really helpful  , it also makes me think there may have been a TB change.
I have not had a chance to test yet.
Christopher

**Comment by cleidigh** [comment]
@kiki-ja 
Okay I just did some quick tests:
- my 78.10 works correctly for both 2.0.4 & .5
- it appears that default setting on my installation enables the font change
do you remember if you changed the default at some point?
I wonder if The Japanese installation might have a different default?
I think this just needs a note describing the requirement.
Christopher

**Comment by kiki-ja** [comment]
@cleidigh 

The default value for 'Allow messages to use other fonts' in Thunderbird seems to be check-in(checked).
It is supposed to be the same for all locales.

In 'about:config' it is as follows.
browser.display.use_document_fonts = 1 (default)
browser.display.use_document_fonts = N/A (none)

I believe that if PrintingTools NG is affected by the current specification, then a description of the requirements will be needed.

[My case]
I use it in check-out(unchecked).
The reason is that I want to view and print in a font that I like.

**Comment by tidebre67** [comment]
@cleidigh 
vote for #5
Thomas

**Comment by cleidigh** [comment]
@tidebre67 /all
has reported the background color option not working.  I think it may be some system option
possibly the colors section. the other observations?
![image](https://user-images.githubusercontent.com/25272315/118375381-46572880-b58f-11eb-8ea9-af9660ff7f87.png)

**Comment by tidebre67** [comment]
@cleidigh : Christopher, my setting are as shown in yr screenprint above. Thomas

**Comment by RRoenn** [comment]
Hello @cleidigh 

I think the best option will be
1. ask the user for permission to change the option (the addon will do the change)

Or 2. silently toggle the system option to match the extension option (because the user want to change the font)

**Comment by RRoenn** [comment]
> @tidebre67 /all
> has reported the background color option not working. I think it may be some system option
> possibly the colors section. the other observations?

Yes, I had also this problem in the beginning.

You had to check this option
![Background Color](https://user-images.githubusercontent.com/77779094/118376092-2a22ae00-b5c6-11eb-8798-fd64544a9c7a.PNG)


You can also applay one of this solution
1.   do nothing since the default settings work without the user doing anything
2.    silently toggle the system option to match the extension option
3.    tell the user to change the system option if not set to allow different fonts
4.    ask the user for permission to change the option
5.    tell the user the system option change (I will prototype this one)
6.    Place a note below the extension option indicating if the system option needs

**Comment by tidebre67** [comment]
@RRoenn : Tks for this info. Works with this settings.
@cleidigh : Christopher:  would propose to make a remark for other users in this respect.

**Comment by cleidigh** [comment]
@RRoenn duh, I always forget that system option! .good  Catch 🥇 
@tidebre67 I see this case as better for a note.
I have a prototype for the pop-up notification on the Font
 pushing right now

**Comment by cleidigh** [comment]
https://github.com/cleidigh/printing-tools-ng/blob/v2.0.6/xpi/printing-tools-ng-2.0.6-b1-tb.xpi

**Comment by RRoenn** [comment]
I am glad that I could help you.

with 6-b1
Yes I see the notification, but as a poor user, I do not know where to change the system option.
So I prefer **4. ask the user for permission to change the option** and the addon will do the change

And a better translation for this option is "Schriftart" (font type)
So please change "Schriftgröße anwenden" to "Schriftart"

A better translation for the last option background color is "Hintergrundfarbe der Kopfzeilen"

A better translation for the wrap attachments per line "Wickeln" is "Umbrechen"

Take your time!

**Comment by cleidigh** [comment]
@RRoenn 
I this is a language issue as the English states that the extension important is changing the system option which I think is better than trying to tell the user to do it
maybe give me also a good translation for the alert in German
```
alert("The system option:\n  Allow messages to use other fonts\nhas been enabled");
		prefs.setIntPref("browser.display.use_document_fonts", 1);
	} else if(notify) {
		alert("The system option:\n  Allow messages to use other fonts\nhas been disabled");
		prefs.setIntPref("browser.display.use_document_fonts", 0);
	}
```

I will also make the other localization changes you suggest, thanks!

**Comment by RRoenn** [comment]
@cleidigh 

```
alert("The system option:\n Allow messages to use other fonts\nhas been enabled");
prefs.setIntPref("browser.display.use_document_fonts", 1);
} else if(notify) {
alert("The system option:\n Allow messages to use other fonts\nhas been disabled");
prefs.setIntPref("browser.display.use_document_fonts", 0);
}
```

This option should only be changed if necessary.
And if `browser.display.use_document_fonts == 1` don't alert if set to 1.
Only alert, if option changed.
If unchecked in the addon, restore user setting ... 1 or 0

German:
1: alert("Systemeinstellung:\n Schriftartenkontrolle: Das Verwenden eigener Schriftarten\nwurde aktiviert");

0: alert("Systemeinstellung:\n Schriftartenkontrolle: Das Verwenden eigener Schriftarten\nwurde deaktiviert");

restore: alert("Systemeinstellung:\n Schriftartenkontrolle: Das Verwenden eigener Schriftarten\nwurde zurückgesetzt");

**Comment by kiki-ja** [comment]
@cleidigh 
and All.

```
options I have thought of:

    1. do nothing since the default settings work without the user doing anything
    2. silently toggle the system option to match the extension option
    3. tell the user to change the system option if not set to allow different fonts
    4. ask the user for permission to change the option
    5. tell the user the system option change (I will prototype this one)
    6. Place a note below the extension option indicating if the system option needs to be set
```

[About approach to options]
My recommendation.
**"2. silently toggle the system option to match the extension option"**

However, it is conditional.
The 'toggle' should _automatically restore_ the original user settings of Thunderbird.
In other words, enable font change only for print preview and print out, and then _automatically restore_ it.
Only when 'browser.display.use_document_fonts' = '0'.
And _add a note or description_ to the options panel in PrintingTools NG.

[The reasons]
・It is not advisable to leave the user settings of Thunderbird changed.
・This allows Thunderbird users to feel safe without being aware of it.
・Thunderbird users can view messages in the font of their choice by language.

**Comment by cleidigh** [comment]
I think we both agree that changing systems settings should not be done 
silently

unless it can be done without issue the idea of restoring the option 
silently also

solves this.  I have to see if there is a way to get a completion event

so I can change it.

This should not be so complicated!

Christopher

**Comment by kiki-ja** [comment]
@cleidigh


> I think we both agree that changing systems settings should not be done silently.

Yes, I agree with you.
Thank you very much.

Warnings and notifications are helpful, but I would guess that many users do not understand them.
It took me a while to research and confirm this issue as well.

I think it would be reasonable to do the option changes and restores automatically and silently.
However, I am concerned about security issues and whether this is technically possible.

**Comment by cleidigh** [comment]
K

I think after some thought and investigation the current method of 
informing the user . Extension

is changing the system option (implemented in b1) is the appropriate way 
to go.  I am open to more suggestions

however I do think trying to implement a silent restore main be 
implementable

more importantly there are some edge conditions that can really cause 
problems

consider canceling after a preview, that is another potential sequence 
requiring interception.

if we have a good dialogue/alert with the system option name that we 
change I think that works.

I just cannot get into something so complicated for this simple option.

the whole printing engine basically have no documentation so this is why 
simple things are often not!

  Christopher

**Comment by cleidigh** [comment]
b2 should be a release candidate
it is fully localized, including @RRoenn suggestions
this version continues to inform you that it will
change the system option when necessary
there is also one improvement from #67 
that is an option to show signatures#vcf attachments

https://github.com/cleidigh/printing-tools-ng/blob/v2.0.6/xpi/printing-tools-ng-2.0.6-b2-tb.xpi

**Comment by cleidigh** [comment]
Fixed with v2.0.6
